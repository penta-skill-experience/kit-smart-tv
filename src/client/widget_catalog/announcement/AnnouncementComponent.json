{
  "REFRESH_RATE": 3000,
  "DISPLAYED_ANNOUNCEMENTS": 5
}