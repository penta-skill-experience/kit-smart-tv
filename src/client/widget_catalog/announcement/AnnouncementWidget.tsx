import {Widget} from "../../widget/Widget";
import {WidgetConfigSaver} from "../../widget/WidgetConfigSaver";

export class AnnouncementWidget implements Widget {

    createConfigComponent(rawConfig: Object, saver: WidgetConfigSaver): JSX.Element {
        return undefined;
    }

    createDisplayComponent(rawConfig: Object): JSX.Element {
        return undefined;
    }

    getTitle(): string {
        return "Announcements";
    }

    isConfigurable(): boolean {
        return false;
    }

}