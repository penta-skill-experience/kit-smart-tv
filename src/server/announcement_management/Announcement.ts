import * as AnnouncementConfig from "./AnnouncementConfig.json";

export class Announcement {

    static AnnouncementAuthorError = class extends Error {

        static DEFAULT_ANNOUNCEMENT_AUTHOR_ERROR_MESSAGE = "Error, the announcement author must be a valid e-mail address";

        constructor(msg: string) {
            super(msg);
        }
    }
    private readonly _title: string;
    private readonly _author: string
    private readonly timeOfAddition: number;

    /**
     * Creates an announcement with the specified parameters.
     * @param title The title of the announcement.
     * @param author The author of the announcement. Must be a valid e-mail address.
     * @param text The text of the announcement.
     * @param timeOfAddition The timeOfAddition. This is optional. If this is not given, the current time is used
     * @param DurationToTimeout The duration until the timeout of this announcement occurs. This is optional. If this is not given a default duration is used.
     *
     * @throws AnnouncementAuthorError gets thrown, if the author is not a valid e-mail address
     */
    public constructor(title: string, author: string, text: string, timeOfAddition?: number, DurationToTimeout?: number) {
        if (!Announcement.validateAuthor(author)) {
            throw new Announcement.AnnouncementAuthorError(Announcement.AnnouncementAuthorError.DEFAULT_ANNOUNCEMENT_AUTHOR_ERROR_MESSAGE);
        }
        this._title = title;
        this._author = author;
        this._text = text;

        this.timeOfAddition = Date.now();
        this._timeout = this.timeOfAddition + +AnnouncementConfig.DEFAULT_ANNOUNCEMENT_TIMEOUT
    }

    private _text: string;

    get text(): string {
        return this._text;
    }

    set text(value: string) {
        this._text = value;
    }

    private _timeout: number;

    get timeout(): number {
        return this._timeout;
    }

    get author(): string {
        return this._author;
    }

    get title(): string {
        return this._title;
    }

// author must be an email-address
    private static validateAuthor(authorToValidate: String): Boolean {
        const validator = require("email-validator");
        return validator.validate(authorToValidate);
    }
}