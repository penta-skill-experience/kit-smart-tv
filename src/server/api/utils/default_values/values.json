{
  "fontSizes": [
    {
      "id": "small",
      "name": "small",
      "relativeSize": 0.8
    },
    {
      "id": "medium",
      "name": "medium",
      "relativeSize": 1.0
    },
    {
      "id": "large",
      "name": "large",
      "relativeSize": 1.15
    }
  ],
  "colorSchemes": [
    {
      "id": "dark",
      "name": "dark",
      "titleFontColor": "white",
      "bodyFontColor": "white",
      "specialBoldFontColor": "ForestGreen",
      "specialSubtleFontColor": "Tomato",
      "accentBarColor": "rgba(0, 0, 0, 0.5)",
      "backgrounds": [
        "https://images.pexels.com/photos/1809644/pexels-photo-1809644.jpeg",
        "https://images.pexels.com/photos/847402/pexels-photo-847402.jpeg?auto=compress&cs=tinysrgb&dpr=2&h=750&w=1260",
        "https://images.pexels.com/photos/2162442/pexels-photo-2162442.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=2",
        "https://images.pexels.com/photos/4328298/pexels-photo-4328298.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=2",
        "https://images.pexels.com/photos/2085998/pexels-photo-2085998.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=2",
        "https://wallpaperaccess.com/full/1379469.jpg",
        "https://images.pexels.com/photos/8824641/pexels-photo-8824641.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=2",
        "https://images.pexels.com/photos/9660579/pexels-photo-9660579.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=2"
      ]
    },
    {
      "id": "light",
      "name": "light",
      "titleFontColor": "black",
      "bodyFontColor": "black",
      "specialBoldFontColor": "green",
      "specialSubtleFontColor": "FireBrick",
      "accentBarColor": "rgba(255,255,255,0.5)",
      "backgrounds": [
        "https://images.pexels.com/photos/8849653/pexels-photo-8849653.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=2",
        "https://images.pexels.com/photos/1323550/pexels-photo-1323550.jpeg?auto=compress&cs=tinysrgb&dpr=2&h=750&w=1260",
        "https://images.pexels.com/photos/691668/pexels-photo-691668.jpeg",
        "https://images.pexels.com/photos/1367192/pexels-photo-1367192.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=2",
        "https://images.pexels.com/photos/4762392/pexels-photo-4762392.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=2",
        "https://images.pexels.com/photos/9716295/pexels-photo-9716295.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=2",
        "https://images.pexels.com/photos/2101187/pexels-photo-2101187.jpeg?auto=compress&cs=tinysrgb&dpr=2&h=750&w=1260",
        "https://images.pexels.com/photos/911738/pexels-photo-911738.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=2"
      ]
    }
  ]
}