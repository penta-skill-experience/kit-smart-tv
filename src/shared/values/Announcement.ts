/**
 * Objects of this interface represent announcements.
 */
export interface Announcement {

    /**
     * An announcement is identified by its title.
     */
    title: string;

    /**
     * The author of an announcement must be an e-mail address.
     */
    author: string;
    text: string;

    /**
     * unix timestamp of the time, when the announcement was added to the system.
     */
    timeOfAddition: number;

    /**
     * unix timestamp of the time, when the timeout of the announcement occurs.
     */
    timeout: number;
}