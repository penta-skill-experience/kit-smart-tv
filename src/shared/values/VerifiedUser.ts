import * as emailValidator from "email-validator"
import {VerifiedUserError} from "./VerifiedUserError";

/**
 * Represents a verified user. A verified user has a name and an e-mail address.
 */
export class VerifiedUser {

    private static readonly INVALID_EMAIL_ERROR_MSG = "Error, the given email was not a valid email-address."

    private _email: string;
    private _name: string;

    /**
     * Creates a verified user with the given email and the given name
     * @param email the given email
     * @param name  the given name
     *
     * @throws VerifiedUserError if the given email is not a valid email
     */
    constructor(email: string, name: string) {
        if (!emailValidator.validate(email)) {
            throw new VerifiedUserError(VerifiedUser.INVALID_EMAIL_ERROR_MSG);
        }
        this._email = email;
        this._name = name;
    }

    get email(): string {
        return this._email;
    }

    get name(): string {
        return this._name;
    }
}
