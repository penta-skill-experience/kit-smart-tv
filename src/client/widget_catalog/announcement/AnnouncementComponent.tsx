import * as React from "react";
import {Announcement} from "../../../server/announcement_management/Announcement";
import {AnnouncementPersistence} from "../../../shared/persistence/AnnouncementPersistence";
import * as AnnouncementConfig from "./AnnouncementComponent.json"

interface AnnouncementState {
    announcements : Announcement[];
}

export class AnnouncementComponent extends React.Component<{}, AnnouncementState> {

    constructor(props) {
        super(props);
        this.state = {
            announcements: []
        };
    }

    async tick() {
        let currentAnnouncements : Announcement[];
        await new AnnouncementPersistence().getAnnouncements().then(data => currentAnnouncements = data);
        this.setState({
            announcements: currentAnnouncements
        });
    }

    async componentWillMount() {
        await this.tick();
    }

    ComponentDidMount() {
        setInterval(() => this.tick(), AnnouncementConfig.REFRESH_RATE);
    }

    render() {
        if (this.state.announcements.length > 0) {
            return this.state.announcements.map(announcement => {
                return <div>
                    <b>{announcement.title} - {announcement.author} </b> <br />
                        {announcement.text} <br />
                    </div>
            })
        } else {
            return <div />
        }
    }
}