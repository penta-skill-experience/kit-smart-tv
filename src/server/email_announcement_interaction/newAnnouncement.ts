import * as AnnouncementConfig from "../announcement_management/AnnouncementConfig.json";
import {AnnouncementAuthorError} from "../announcement_management/AnnouncementAuthorError";
import * as emailValidator from "email-validator";
import {Announcement} from "../../shared/values/Announcement";

/**
 * Creates an announcement with the specified parameters.
 * @param title The title of the announcement.
 * @param author The author of the announcement. Must be a valid e-mail address.
 * @param text The text of the announcement.
 * @param timeOfAddition The timeOfAddition. This is optional. If this is not given, the current time is used
 * @param durationToTimeout The duration until the timeout of this announcement occurs. This is optional. If this is not given a default duration is used.
 *
 * @throws AnnouncementAuthorError gets thrown, if the author is not a valid e-mail address
 */
export function newAnnouncement(title: string, author: string, text: string,
                                timeOfAddition: number = Date.now(),
                                durationToTimeout: number = AnnouncementConfig.DEFAULT_ANNOUNCEMENT_TIMEOUT): Announcement {

    if (!emailValidator.validate(author)) {
        throw new AnnouncementAuthorError(
            AnnouncementAuthorError.DEFAULT_ANNOUNCEMENT_AUTHOR_ERROR_MESSAGE);
    }
    return {
        title: title,
        author: author,
        text: text,
        timeOfAddition: timeOfAddition,
        timeout: timeOfAddition + durationToTimeout,
    }
}

