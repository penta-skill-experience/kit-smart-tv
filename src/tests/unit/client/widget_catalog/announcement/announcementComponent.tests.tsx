import React from "react";
import renderer from 'react-test-renderer';
import {VerifiedUser} from "../../../../../shared/values/VerifiedUser";
import {newAnnouncement} from "../../../server/util/newAnnouncement";
import {
    AnnouncementPersistenceBackend
} from "../../../../../shared/persistence/announcements/AnnouncementPersistenceBackend";
import {Announcement} from "../../../../../shared/values/Announcement";
import {AnnouncementComponent} from "../../../../../client/widget_catalog/announcement/AnnouncementComponent";

// setting up test values
const bob = new VerifiedUser("bob@example.com", "bob");
const alice = new VerifiedUser("alice@example.com", "alice");
const verifiedUsers = [bob, alice];

const bobAnnouncement = newAnnouncement("Bobs Announcement", bob.email, "This is bobs announcement.");
const aliceAnnouncement = newAnnouncement(
    "Announcement from alice", alice.email, "This announcement is from alice");
const announcements = [bobAnnouncement, aliceAnnouncement];

//mocking announcementPersistence to return test values for these tests
const getAnnouncementsMock = jest.spyOn(AnnouncementPersistenceBackend.prototype, "getAnnouncements");
const getVerifiedUsersMock = jest.spyOn(AnnouncementPersistenceBackend.prototype, "getVerifiedUsers");

getAnnouncementsMock.mockImplementation(() => {
    return new Promise<Announcement[]>(resolve => {
        resolve(announcements);
    });
});
getVerifiedUsersMock.mockImplementation(() => {
    return new Promise<VerifiedUser[]>(resolve => {
        resolve(verifiedUsers);
    });
});

test("announcementComponent Snapshot", () => {
    const tree = renderer
        .create(<AnnouncementComponent error={(msg => {})} specialBoldFontColor={"ForestGreen"} specialSubtleFontColor={"DarkOrange"} />)
            .toJSON();
    expect(tree).toMatchSnapshot();
});

afterAll(() => {
    jest.restoreAllMocks();
})