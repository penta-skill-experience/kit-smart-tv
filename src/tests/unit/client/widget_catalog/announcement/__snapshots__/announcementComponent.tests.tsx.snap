// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`announcementComponent Snapshots announcementComponent Snapshot with announcements 1`] = `
<div
  className="grid grid-flow-row"
>
  <div
    className="font-light leading-normal sm:text-xs lg:text-base xl:text-base 2xl:text-xl 4xl:text-2xl sm:text-left 8xl:text-4xl"
    key="0"
  >
    <b>
      Announcement from alice
       - 
      alice@example.com
       
    </b>
    <br />
    This announcement is from alice
     
    <br />
    <br />
  </div>
  <div
    className="font-light leading-normal sm:text-xs lg:text-base xl:text-base 2xl:text-xl 4xl:text-2xl sm:text-left 8xl:text-4xl"
    key="1"
  >
    <b>
      Bobs Announcement
       - 
      bob
       
    </b>
    <br />
    This is bobs announcement.
     
    <br />
  </div>
</div>
`;

exports[`announcementComponent Snapshots announcementComponent Snapshot without announcements 1`] = `
<div
  className="grid grid-flow-row"
/>
`;
